include/envoy/http/filter.h

```diff
@@ -532,11 +532,27 @@ class FilterChainFactory {
   virtual ~FilterChainFactory() {}
 
   /**
-   * Called when a new stream is created on the connection.
+   * Called when a new HTTP stream is created on the connection.
    * @param callbacks supplies the "sink" that is used for actually creating the filter chain. @see
    *                  FilterChainFactoryCallbacks.
    */
   virtual void createFilterChain(FilterChainFactoryCallbacks& callbacks) PURE;
+
+  /**
+   * Called when a new upgrade stream is created on the connection.
+   * @param upgrade supplies the upgrade header from downstream
+   * @param callbacks supplies the "sink" that is used for actually creating the filter chain. @see
+   *                  FilterChainFactoryCallbacks.
+   */
+  virtual bool createUpgradeFilterChain(absl::string_view upgrade,
+                                        FilterChainFactoryCallbacks& callbacks) PURE;
+
+  /**
+   * Called to determine if this type of upgrade is supported.
+   * @param the upgrade type requested.
+   * @return true if upgrades of this type are allowed, false otherwise.
+   */
+  virtual bool upgradeSupported(absl::string_view upgrade) const PURE;
```

Comment:
Where is the called from in prod code? Is the intention that a filter say at runtime that it doesn't support an upgrade? That scares me a little bit just from a config stability perspective, but mainly wondering how this is intended to work. I assume this is going to be used in a follow up? Maybe more docs here?

Reply:
It's currently used in the HCM where we decide if we're accepting or rejecting upgrade headers.   That said we call createFilterChain before we make that go no-go call, so I could latch the return value of createUpgradeFilterChain, return that from createFilterChain and use that for the reject instead of calling isUpgradeSupported

https://github.com/envoyproxy/envoy/compare/master...alyssawilk:websocket_delay#diff-d70ea2d9706e0246700148b2e2bb63ceR569 (doesn't have error handling yet)